Add support for QueryBuilder upsert() method.

LukeTowers opened the pull request:
Pulls in the work by @paras-malhotra in https://github.com/laravel/framework/pull/34698 & https://github.com/laravel/framework/pull/34712 for use in October CMS.

Commit:
Add support for QueryBuilder upsert() method.

Pulls in the work by @paras-malhotra in https://github.com/laravel/framework/pull/34698 & https://github.com/laravel/framework/pull/34712 for use in October CMS.

LukeTowers commented:
@bennothommo if you could adapt the unit tests from those two PRs for this library, that would be fantastic :)

bennothommo reviewed (approved):
I like it! Everything looks good to me.

bennothommo commented:
@LukeTowers I shall port the unit tests shortly :)

Commit:
Add unit tests for upsert() query builder method.

Based off implementation done in https://github.com/laravel/framework/pull/34698 and https://github.com/laravel/framework/pull/34712.

Commit:
Drop separate selectConcat() method tests

bennothommo commented:
@LukeTowers unit tests added.

mjauvin commented on tests/Database/QueryBuilderTest.php:
```diff
@@ -0,0 +1,259 @@
+        return $connection;
+    }
+
+    protected function getBuilder()
```
@LukeTowers where is this method used?

LukeTowers replied:
Not sure, ask @bennothommo

bennothommo replied:
@mjauvin It's not used, but it may be in the future if we add more generic tests for the builder. I ported these methods from Laravel, to be fair.
